### PR TITLE
Allow query string parameters without values

### PR DESCRIPTION
**Pull request by moozzyk**
Addresses #624

### CODE REVIEW

**Comment by khellang** [comment]
LGTM :shipit:

File: test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs [hunk 1]
```diff
@@ -24,5 +24,38 @@ public void QueryReturnsParsedQueryCollection()
             // Assert
             Assert.Equal("bar", queryCollection["foo"]);
         }
+
+        [Theory]
+        [InlineData("?q", "q")]
+        [InlineData("?q&", "q")]
+        [InlineData("?q1=abc&q2", "q2")]
+        public void QueryCollection_TEST(string queryString, string emptyParam)
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs]: Test names need work. Component_Scenario

File: test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs [hunk 2]
```diff
@@ -24,5 +24,38 @@ public void QueryReturnsParsedQueryCollection()
...
+            var features = new FeatureCollection();
+            var request = new HttpRequestFeature();
+            request.QueryString = queryString;
+            features[typeof(IHttpRequestFeature)] = request;
+
+            var provider = new QueryFeature(features);
+
+            var queryCollection = provider.Query;
+
+            Assert.Equal(string.Empty, queryCollection[emptyParam]);
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs]: This was true before. Check queryCollection.Keys.Contains?

File: test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs (hunk 1 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs]: My bad - forgot to rename.

File: test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs (hunk 2 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs]: Will check both.

**Comment by moozzyk** [comment]
:up:date: - For the tests I used the same convention as used by the test that already existed.

File: test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs [hunk 3]
```diff
@@ -24,5 +24,39 @@ public void QueryReturnsParsedQueryCollection()
...
+            Assert.True(queryCollection.Keys.Contains(emptyParam));
+            Assert.Equal(string.Empty, queryCollection[emptyParam]);
+        }
+
+        [Fact]
+        public void EmptyKeysNotAddedToQueryCollection()
+        {
+            var features = new FeatureCollection();
+            var request = new HttpRequestFeature();
+            request.QueryString = "&&";
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs]: there should be a leading `?` on here.

File: test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs [hunk 4]
```diff
@@ -24,5 +24,39 @@ public void QueryReturnsParsedQueryCollection()
             // Assert
             Assert.Equal("bar", queryCollection["foo"]);
         }
+
+        [Theory]
+        [InlineData("?q", "q")]
+        [InlineData("?q&", "q")]
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs]: Add `?q=` and `?q=&`

**Comment by Tratcher** [comment]
:shipit:

File: test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs (hunk 3 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs]: I did not put the leading '?' based on https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.WebUtilities/QueryHelpers.cs#L126 (and the code that follows this). I don't feel strongly about this so I can add '?' if you prefer to have it.

Comment (moozzyk) [inline test/Microsoft.AspNetCore.Http.Tests/Features/QueryFeatureTests.cs]: I will do this: 
```
        [Theory]
        [InlineData("?&&")]
        [InlineData("?&")]
        [InlineData("&&")]
```

**Pull request merged**